File: rpi/service/car_service.py
from car.car import Car
from service.car_thread import CarThread
from service.image_analysis import ImageAnalysisService

class CarService:
    def __init__(self, car: Car, image_analysis_service: ImageAnalysisService):
        self._car = car
        self._imageAnalysisService = image_analysis_service
        self._thread = None

    def run(self):
        if self._thread:
            return

        self._thread = CarThread(car=self._car, image_analysis_service=self._imageAnalysisService)
        self._thread.start()

    def stop(self):
        if self._thread:
            self._thread.stop()
            self._thread.join()
            self._thread = None

File: client/src/app/car/car.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import * as io from 'socket.io-client';
import { CarStatus, EventModel } from './car.model';

@Injectable({
  providedIn: 'root',
})
export class CarService {
  socket: SocketIOClient.Socket;
  events: EventModel[] = [];

  constructor(private http: HttpClient) {}

  connect(): void {
    console.log('CarService => connect');
    const socket = io.connect('http://localhost:5000');
    this.socket = socket;

    socket.on('connect', () => {
      console.log('SCOKET => connect');
      socket.send('Client connected');
    });

    socket.on(CarStatus.TAKING_IMAGE, (data) => {
      console.log('SOCKET =>', CarStatus.TAKING_IMAGE);
      const event: EventModel = {
        type: CarStatus.TAKING_IMAGE,
        data: data,
      };
      this.events.push(event);
    });

    socket.on(CarStatus.IMAGE_TAKEN, (data) => {
      console.log('SOCKET =>', CarStatus.IMAGE_TAKEN);
      const event: EventModel = {
        type: CarStatus.IMAGE_TAKEN,
        data: data,
      };
      this.events.push(event);
    });
  }

  start() {
    console.log('CarService => startCar');
    this.http.put('http://localhost:5000/run', {}).subscribe();
  }

  stop() {
    console.log('CarService => stopCar');
    this.http.post('http://localhost:5000/stop', {}).subscribe();
  }

  disconnect() {
    console.log('CarService => disconnect');
    this.socket.disconnect();
  }
}
